core/trino-main/src/main/java/io/trino/sql/analyzer/JsonPathAnalyzer.java

```diff
@@ -0,0 +1,533 @@
+import io.trino.sql.jsonpath.tree.ArithmeticBinary;
+import io.trino.sql.jsonpath.tree.ArithmeticUnary;
+import io.trino.sql.jsonpath.tree.ArrayAccessor;
+import io.trino.sql.jsonpath.tree.ArrayAccessor.Subscript;
+import io.trino.sql.jsonpath.tree.CeilingMethod;
+import io.trino.sql.jsonpath.tree.ComparisonPredicate;
+import io.trino.sql.jsonpath.tree.ConjunctionPredicate;
+import io.trino.sql.jsonpath.tree.ContextVariable;
+import io.trino.sql.jsonpath.tree.DatetimeMethod;
+import io.trino.sql.jsonpath.tree.DisjunctionPredicate;
+import io.trino.sql.jsonpath.tree.DoubleMethod;
+import io.trino.sql.jsonpath.tree.ExistsPredicate;
+import io.trino.sql.jsonpath.tree.Filter;
+import io.trino.sql.jsonpath.tree.FloorMethod;
+import io.trino.sql.jsonpath.tree.IsUnknownPredicate;
+import io.trino.sql.jsonpath.tree.JsonNullLiteral;
+import io.trino.sql.jsonpath.tree.JsonPath;
+import io.trino.sql.jsonpath.tree.JsonPathTreeVisitor;
+import io.trino.sql.jsonpath.tree.KeyValueMethod;
+import io.trino.sql.jsonpath.tree.LastIndexVariable;
+import io.trino.sql.jsonpath.tree.LikeRegexPredicate;
+import io.trino.sql.jsonpath.tree.MemberAccessor;
+import io.trino.sql.jsonpath.tree.NamedVariable;
+import io.trino.sql.jsonpath.tree.NegationPredicate;
+import io.trino.sql.jsonpath.tree.PathNode;
+import io.trino.sql.jsonpath.tree.PredicateCurrentItemVariable;
+import io.trino.sql.jsonpath.tree.SizeMethod;
+import io.trino.sql.jsonpath.tree.SqlValueLiteral;
+import io.trino.sql.jsonpath.tree.StartsWithPredicate;
+import io.trino.sql.jsonpath.tree.TypeMethod;
+import io.trino.sql.tree.Node;
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.StringLiteral;
+
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.spi.StandardErrorCode.INVALID_PATH;
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static io.trino.spi.function.OperatorType.NEGATION;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.VarcharType.createVarcharType;
+import static io.trino.sql.analyzer.ExpressionAnalyzer.isCharacterStringType;
+import static io.trino.sql.analyzer.ExpressionAnalyzer.isNumericType;
+import static io.trino.sql.analyzer.ExpressionAnalyzer.isStringType;
+import static io.trino.sql.analyzer.ExpressionTreeUtils.extractLocation;
+import static io.trino.sql.analyzer.SemanticExceptions.semanticException;
+import static io.trino.sql.analyzer.TypeSignatureProvider.fromTypes;
+import static io.trino.sql.jsonpath.tree.ArithmeticUnary.Sign.PLUS;
+import static io.trino.type.Json2016Type.JSON_2016;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class JsonPathAnalyzer
```

Comment:
what are the rules regarding expressions within json path (coercions, evaluation order, ...) vs expression in SQL. Are those the same?

Reply:
The path expressions which "resemble" SQL expressions, are generally supposed to have the same semantics as SQL. Coercions and evaluation order are the same.
The main differences are the following:
- path expressions process sequences of input
- error handling in path is specified by lax / strict mode + the ON ERROR clause
- boolean expressions in path filter have their own error handling rules and special semantics. So, in the path filter the operators like `<`, `>`,... behave different that their SQL counterparts (see the javadoc in `JsonPredicateEvaluator` class).